### ISSUE TITLE
Enhance `foreach` loops to allow iterating over multiple, parallel collections in tandem (pairs)

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#14732]
## Summary of the new feature/enhancement

Iterating over multiple collections that have corresponding elements iteratively is a common scenario.
This is currently fairly cumbersome, so letting the `foreach` statement provide syntactic sugar would be a nice simplification (and would amount to a generalization of the pairwise [`Linq.Enumerable.Zip()`](https://docs.microsoft.com/en-US/dotnet/api/System.Linq.Enumerable.Zip) method):

```powershell
# 3 collections with corresponding elements('foo' relates to 'baz' and 'and', 'bar' relates to 'qux' and 'so on')
$a = 'foo', 'bar'
$b = 'baz', 'qux'
$c = 'and', 'so on'

# WISHFUL THINKING
# Iterate over the corresponding elements from each RHS collection.
# Note: Number of iterator variables must match the number of collections.
foreach ($aElem, $bElem, $cElem in $a, $b, $c) { 
   "$aElem, $bElem, $cElem"
}
```

The above would yield:

```none
foo, baz, and
bar, qux, so on
```

and would be the equivalent of the (far more cumbersome) following:

```powershell
foreach ($i in 0..($a.count-1)) {
  $aElem, $bElem, $cElem = $a[$i], $b[$i], $c[$i]
  "$aElem, $bElem, $cElem"
}
```

Due to its resemblance to destructuring assignments, I think the purpose of the syntax is easy to infer.

Note:

* With _multiple_ iterator variables on the LHS  (which activates the new feature being introduced), their number would be required to match the number of explicitly listed collections (RHS); e.g.:
  * `foreach ($aElem, $bElem, $cElem in $a, $b, $c) { ... }` ... VALID: 3 iterator variables, 3 collections
  * `foreach ($aElem, $bElem in $a, $b, $c) { ... }`  ... INVALID: only 2 iterator variables vs. 3 collections
 
   * The _single_-iterator variable (the currently supported syntax) would continue to function as-is; e.g.:

     * `foreach ($elem in $a, $b, $c) { ... }` ... RHS is conceived as a *single* collection to iterate over directly, resulting in 3 iterations with `$elem` bound to the values of `$a`, `$b`, `$c` in sequence (irrespective of whether these variables contain scalars or collections).
 
* The _largest_ among the RHS collections would drive the number of iterations, and the iterator variables would contain `$null` for those collections that have run out of elements.

Backward-compatibility considerations: N/A, because the proposed new syntax currently results in a syntax error.

**Comment by iRon7** [comment]
As this is already a valid syntax:

```PowerShell
$aElem, $bElem = 'One', 1
```

I would expect that the iterator part (`$aElem, $bElem` as in the example below) should at already work (but unfortunately doesn't):

```PowerShell
[string[]] $a = 'one', 'two', 'three'
[int[]] $b = 1, 2, 3

foreach ($aElem, $bElem in [Linq.Enumerable]::Zip($a1, $a2)) {
  '{0}: {1}' -f $aElem, $bElem
}
```

**Comment by iRon7** [comment]
Hmmm, that also would imply that

```
[string[]] $a = 'one', 'two', 'three'
[int[]] $b = 1, 2, 3


foreach ($aElem in [Linq.Enumerable]::Zip($a1, $a2)) { $aElem }
```
(Where `$bElem` is left out)

Should just return:
```
one
two
three
```
And not a *tuple*.
on closer inspection, I think this purpose is very confusing...

**Comment by mklement0** [comment]
@iRon7, the `in` in a `foreach` statement is _not_ an assignment, it is a distinct syntax construct that can be thought of as a higher-level assignment:

It embodies an _instruction_ for a (behind-the-scenes) assignment to the iterator variable in every loop iteration, namely to the current element of the enumeration.

As such, there's no inherent requirement to support _multiple_ iterator variables, the way true assignments support multiple target variables in destructuring assignments.

What I meant to express in the OP is that allowing multiple iterator variables is _inspired_ by regular destructuring assignments, in that they would function analogously, albeit with _enforced symmetry_:

The number of iterator variables on the LHS would have to match the number of _explicitly listed_ collections on the RHS:

That is, the following would be syntactically valid:

```powershell
$a = 'foo', 'bar'
$b = 'baz', 'qux'

# OK: two iterator variables, 2 collections.
foreach ($aElem, $bElem in $a, $b) { 
   "$aElem, $bElem"
}

# Ditto
foreach ($aElem, $bElem in (Get-ChildItem foo/), (Get-ChildItem bar/)) { 
   "$aElem, $bElem"
}
```

By contrast, `foreach ($aElem, $bElem in $Pairs)` would _not_ be valid, because there's only _one_ RHS collection for the 2 LHS iterator variables.

**Comment by bpayette** [comment]
@mklement0 Quick note: your proposed semantic for the LHS of `in` would collide with the currently defined behaviour:
```powershell
PS> foreach ($i in 1,2,3) {$i}
1
2
3
```
(@iRon7 's suggestion would work and it's what I would have implemented if I'd had time since it naturally falls out of destructuring.)

That said,  the `foreach` statement could accept additional tokens besides `in` such as `over` or `from`.  Or something after the `foreach` e.g. `foreach pair ($x, $y in $list1,$list2) { ... }`. Basically the syntax is up for grabs - there is no need to try to retrofit new semantics over old syntax.

**Comment by mklement0** [comment]
Thanks, @BrucePay, but I don't think there would be a collision, at least not _technically_:

* `foreach ($i in 1,2,3) { $i }` - i.e. the _single_-iterator case - would continue to function as before: the RHS operand is a _single, flat collection_ (or flattened on demand) to whose elements the iterator variable is bound in sequence.

* Only _multiple_ iterator variables trigger the new behavior, in which case the number of iterator variables must then match the number of _explicitly listed multiple collections_ on the RHS; to use a simplified example with array literals (in the real world I would expect the collections to be stored in variables or to come from commands):

  * `foreach ( $i, $j in (1 ,2, 3), ('a', 'b', 'c') ) { $i, $j }` (in the first iteration, `$i` is `1` and `$j` is `'a'`, ...)

I've updated the OP to make that clearer.

If the consensus ends up being that this distinction is too subtle, a new syntax could be pondered, but personally I don't think it's necessary.

**Comment by iRon7** [comment]
I needed some time to think this through and enhance my [`Join-Object`](https://www.powershellgallery.com/packages/Join) cmdlet (without losing its existing features), but I think that something similar is possible with a "`Join-Object`" cmdlet by building a list of (`[Collections.ObjectModel.Collection[psobject]]`) collections that don't immediately destruct themselves (but will when they e.g. being assigned to a single variable).
Meaning that without [*investing too much in traditional operators*](https://github.com/PowerShell/PowerShell/issues/14724#issuecomment-777526400), you might investigate in an standard `Join-Object` cmdlet instead:

```PowerShell
$a = 'foo', 'bar'
$b = 'baz', @(1,2)
$c = 'and', 'so on'

$a |Join $b |Join $c |% {
    $aElem, $bElem, $cElem = $_
    "$aElem | $bElem | $cElem"
}

foo | baz | and
bar | 1 2 | so on
```

See: [`Join-Object`](https://github.com/iRon7/Join-Object) issue: [`#14` Support non-object arrays](https://github.com/iRon7/Join-Object/issues/14) for more examples.
And my [Add a Join-Object cmdlet to the standard PowerShell equipment]() purpose.

**Comment by mklement0** [comment]
Re not "investing too much in traditional operators" (read: statements).
This strikes me as a _false dichotomy_, as I've argued in https://github.com/PowerShell/PowerShell/issues/14724#issuecomment-777801399.

Therefore, this is again not an either-or scenario, so I suggest you add your previous example to #14994

**Comment by iRon7** [comment]
> This strikes me as a _false dichotomy_

Agree, I have rephrased it in the comment of this issue and removed it from the [`Join-Object` purpose]()](https://github.com/PowerShell/PowerShell/issues/14994)

**Comment by iRon7** [comment]
Thinking outside the box and following PowerShell's streaming concept.
It would be nice if we could have parallel input streams rather than pushing for multiple in-memory arrays.

This:
```PowerShell
$x = (1..9 |)
```
Causes currently an error:
> An empty pipe element is not allowed.

Instead it would be nice if it would create a kind of a "**deferred pipeline object**" instead and each time the object is invoked it processes and returns the next item in the deferred parallel pipeline (until it is empty where it returns a `AutomationNull`)

Wishful thinking:
```PowerShell
$a = (1..9 |) # For the example below this `$a = 1..9` would have the results but (presumably uses more memory)
$b = (Get-Connect .\MyHughFile.txt |) # or any other long stream
$c = (Import-Csv .\Large.csv |)
$a | ForEach-Object {
    Write-Host 'current $a item' $_ 
    Write-Host 'current $b item' $b # Every time the $b is called, it processes the next $b item
    Write-Host 'current $b item' $c 
}
```
A deferred pipeline object would also allow to breakup a lengthy pipeline syntax and assign parts to a variable without loading it into memory.

**Comment by mklement0** [comment]
@iRon7, implementing an _analogous_ feature for use _streaming use_, in the pipeline, makes sense to me (as in the `$PSIndex`, index-variable in `foreach` pairing).

I do suggest creating a separate issue for that.

As a thought up front: I'm not sure we need new syntax for that, perhaps even a `ForEach-Object` enhancement along the following lines is doable, using script blocks _as input_ that are then executed and stepped through in parallel:

```
# WISHFUL THINKING
{ 0..2 }, { 'a'..'c' } | ForEach-Object -InvokeScriptBlocks { '{0}: {1}' -f $_[0], $_[1] }
0: a
1: b
2: c
```

**Comment by rjmholt** [comment]
After discussing this with the Engine working group, we don't think this should be baked in at the language/syntax level. This is probably best accomplished with a function like Python's [enumerate](https://python-reference.readthedocs.io/en/latest/docs/functions/enumerate.html). This could be implemented in an external module and later evaluated for inclusion in PowerShell.

**Comment by mklement0** [comment]
> A function for that in PowerShell could be implemented in an external module first

* @SeeminglyScience has posted a proof-of-concept function, `Invoke-ForEachOfEach`, at https://github.com/PowerShell/PowerShell/issues/20126#issuecomment-1690530464, along with thoughts on why making changes to the language itself is tricky.
Ultimately, however, implementation as part of the language is the only solution that makes sense to me, not least for the sake of performance.
